pkg/sdkstats/sdkstats.go

```diff
@@ -2,161 +2,120 @@ package sdkstats
-
-	if err != nil {
-		return nil, err
-	}
-
-	if strings.Contains(q, "helm") {
-		projectType = "helm"
-	}
-	if strings.Contains(q, "ansible") {
-		projectType = "ansible"
-	}
-	if strings.Contains(q, "go.mod") {
-		projectType = "go.mod"
-	}
-	if strings.Contains(q, "Gopkg.toml") {
-		projectType = "Gopkg.toml"
-	}
-	for j := 0; j < len(searchOp); j++ {
-		for i := 0; i < len(searchOp[j].CodeResults); i++ {
-			owner := searchOp[j].CodeResults[i].GetRepository().GetOwner().GetLogin()
-			name := searchOp[j].CodeResults[i].GetRepository().GetName()
-			fullName := searchOp[j].CodeResults[i].GetRepository().GetFullName()
-
-			//getVersion returns SDK version by doing Fragment search
-			if projectType == "Gopkg.toml" {
-				sdkVersion = 0
-			} else {
-				sdkVersion = getversion(searchOp[j].CodeResults[i], projectType)
-			}
+//RepoMetadataQuery struct to store Search queries, and Project Type
+type RepoMetadataQuery struct {
+	ProjectType   string
+	Queries       []string
+	VersionParser VersionParser
+}
 
-			//GetRepoDetails returns Repository specififc details
-			repoMap := GetRepoDetails(client, owner, name)
+//VersionParser Interface implements ParseVersion, to Parse Version number from a given Text Match result.
+type VersionParser interface {
+	ParseVersion(codeResults github.CodeResult, projectType string) (string, error)
+}
 
-			TotalCommits, err := strconv.Atoi(repoMap["TotalCommits"])
-			Stars, err := strconv.Atoi(repoMap["Stars"])
-			if err != nil {
-				return nil, err
-			}
+//GetStats returns []RepoMetaData populated with Code results from Github Search Code API, for a given search string
+func GetStats(client *github.Client, rq RepoMetadataQuery) ([]RepoMetadata, error) {
+	var repoList []RepoMetadata
+	var sdkVersion string
+	var skipVersion int
 
-			repoDetails := &RepoMetadata{
-				FullName:     fullName,
-				ProjectType:  projectType,
-				SDKVersion:   sdkVersion,
-				Stars:        Stars,
-				CreatedAt:    repoMap["CreatedAt"],
-				PushedAt:     repoMap["PushedAt"],
-				TotalCommits: TotalCommits,
+	for _, q := range rq.Queries {
+		searchOp, err := Search(client, q)
+		if err != nil {
+			return nil, err
+		}
+		if strings.Contains(q, "Gopkg.toml") {
+			skipVersion = 1
+		}
+		for j := 0; j < len(searchOp); j++ {
+			for i := 0; i < len(searchOp[j].CodeResults); i++ {
+				if skipVersion == 1 {
+					sdkVersion = "N/A"
+				} else {
+					sdkVersion, err = rq.VersionParser.ParseVersion(searchOp[j].CodeResults[i], rq.ProjectType)
+					if err != nil {
+						return nil, err
+					}
+				}
+				repoDetails := &RepoMetadata{
+					URL:         searchOp[j].CodeResults[i].GetRepository().GetURL(),
+					Name:        searchOp[j].CodeResults[i].GetRepository().GetName(),
+					Owner:       searchOp[j].CodeResults[i].GetRepository().GetOwner().GetLogin(),
+					ProjectType: rq.ProjectType,
+					SDKVersion:  sdkVersion,
+				}
+				repoDetails, err = GetRepoDetails(client, repoDetails)
```

Comment:
Accounting for the above comment about using `range`, maybe just pass `codeResult` into `GetRepoDetails()`, and have that function create the `RepoMetadata` struct and fill everything in?

Reply:
You mean, Move below code into GetRepoDetails()?
				if skipVersion == 1 {
					sdkVersion = "N/A"
				} else {
					sdkVersion, err = rq.VersionParser.ParseVersion(searchOp[j].CodeResults[i], rq.ProjectType)
					if err != nil {
						return nil, err
					}
				}
				repoDetails := &RepoMetadata{
					URL:         searchOp[j].CodeResults[i].GetRepository().GetURL(),
					Name:        searchOp[j].CodeResults[i].GetRepository().GetName(),
					Owner:       searchOp[j].CodeResults[i].GetRepository().GetOwner().GetLogin(),
					ProjectType: rq.ProjectType,
					SDKVersion:  sdkVersion,
				}
				repoDetails, err = GetRepoDetails(client, repoDetails)
				if err != nil {
					return nil, err
				}
				repoList = append(repoList, *repoDetails)

Reply:
However, see below.
The reason I made these exportable, is that, In case other users have the need to let say only call "Search()", or "GetRepoDetails", It would be simple enough to call them. WDYT. This also makes case, NOT to pass "CodeResult" to GetRepoDetails(). WDYT?

Reply:
IMO its always best to err on the side of caution and wait until you need another function to be exported or a user asks for new functionality.  I often find that my first implementation of something can be improved when I come back to it days, weeks, or months later.

If it's already an exported API, I'm kind of stuck. By leaving it unexported, I can change its function signature, refactor functions and their names, etc.

Reply:
Sounds reasonable, Will change